src/iosrv/drpc_ras.c

```diff
@@ -0,0 +1,319 @@
+	if (eic != SHARED__RASEVENT__EXTENDED_INFO_STR_INFO)
+		goto out;
+	if (evt->str_info != NULL)
+		D_FPRINTF(stream, " data: [%s]", evt->str_info);
+
+out:
+	fclose(stream);
+	D_DEBUG(DB_MGMT, "&&& RAS EVENT%s", buf);
+	D_FREE(buf);
+}
+
+static int
+send_event(Shared__RASEvent *evt)
+{
+	Shared__ClusterEventReq	 req = SHARED__CLUSTER_EVENT_REQ__INIT;
+	uint8_t			*reqb;
+	size_t			 reqb_size;
+	Drpc__Call		*dreq;
+	Drpc__Response		*dresp;
+	int			 rc;
+
+	if (dss_drpc_ctx == NULL) {
+		D_ERROR("dRPC not connected: "DF_RC"\n", DP_RC(-DER_UNINIT));
+		return -DER_UNINIT;
+	}
+
+	if (evt == NULL) {
+		D_ERROR("null RAS event\n");
+		return -DER_INVAL;
+	}
+	req.event = evt;
+
+	reqb_size = shared__cluster_event_req__get_packed_size(&req);
+	D_ALLOC(reqb, reqb_size);
+	if (reqb == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	shared__cluster_event_req__pack(&req, reqb);
+	rc = drpc_call_create(dss_drpc_ctx, DRPC_MODULE_SRV,
+			      DRPC_METHOD_SRV_CLUSTER_EVENT, &dreq);
+	if (rc != 0) {
+		D_FREE(reqb);
+		goto out;
+	}
+
+	dreq->body.len = reqb_size;
+	dreq->body.data = reqb;
+
+	rc = drpc_call(dss_drpc_ctx, R_SYNC, dreq, &dresp);
+	if (rc != 0)
+		goto out_dreq;
+	if (dresp->status != DRPC__STATUS__SUCCESS) {
+		D_ERROR("received erroneous dRPC response: %d\n",
+			dresp->status);
+		rc = -DER_IO;
+	}
+
+	drpc_response_free(dresp);
+
+out_dreq:
+	/* This also frees reqb via dreq->body.data. */
+	drpc_call_free(dreq);
+out:
+	free_event(evt);
+
+	return rc;
+}
+
+static int
+raise_ras(ras_event_t id, char *msg, ras_type_t type, ras_sev_t sev, char *hid,
+	  d_rank_t *rank, char *jid, uuid_t *puuid, uuid_t *cuuid,
+	  daos_obj_id_t *oid, char *cop, Shared__RASEvent *evt)
+{
+	int	rc;
+
+	rc = init_event(id, msg, type, sev, hid, rank, jid, puuid, cuuid, oid,
+			cop, evt);
+	if (rc != 0) {
+		D_ERROR("failed to init RAS event %s: "DF_RC"\n",
+			ras_event2str(RAS_POOL_REPS_UPDATE), DP_RC(rc));
+		return rc;
+	}
+
+	log_event(evt);
+	rc = send_event(evt);
+	if (rc != 0)
+		D_ERROR("failed to send RAS event %s over dRPC: "DF_RC"\n",
+			ras_event2str(RAS_POOL_REPS_UPDATE), DP_RC(rc));
+
+	free_event(evt);
+
+	return rc;
+}
+
+void
+ds_notify_ras_event(ras_event_t id, char *msg, ras_type_t type, ras_sev_t sev,
+		    char *hid, d_rank_t *rank, char *jid, uuid_t *puuid,
+		    uuid_t *cuuid, daos_obj_id_t *oid, char *cop, char *data)
+{
+	Shared__RASEvent	evt = SHARED__RASEVENT__INIT;
+	d_rank_t		this_rank = dss_self_rank();
+
+	/* use opaque blob oneof case for extended info for passthrough event */
+	evt.extended_info_case = SHARED__RASEVENT__EXTENDED_INFO_STR_INFO;
+	evt.str_info = data;
+
+	/* populate rank param if empty */
+	if (rank == NULL)
+		rank = &this_rank;
+
+	raise_ras(id, msg, type, sev, hid, rank, jid, puuid, cuuid, oid, cop,
+		  &evt);
+}
+
+int
+ds_notify_pool_svc_update(uuid_t *puuid, d_rank_list_t *svc)
+{
+	Shared__RASEvent			evt = SHARED__RASEVENT__INIT;
+	Shared__RASEvent__PoolSvcEventInfo	info = \
```

Comment:
(style)  Avoid unnecessary line continuations

Reply:
This one is real.

Reply:
I tried different approaches to avoid this line continuation but due to issues with runtime protobuf constructor not behaving as expected in cmocka unit tests and the length of the proto-generated type names being so long, I will have another go at removing it in the next PR.